File: backend/back/src/main/java/com/back/domain/welfare/policy/entity/Policy.kt
package com.back.domain.welfare.policy.entity

import com.back.domain.welfare.policy.dto.PolicyFetchResponseDto.PolicyItem
import jakarta.persistence.*

@Entity
@Table(name = "policy")
class Policy(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    var id: Int? = null,

    @Column(name = "plcy_no", nullable = false, unique = true)
    var plcyNo: String? = null,

    var plcyNm: String? = null,
    var plcyKywdNm: String? = null,

    @Lob
    var plcyExplnCn: String? = null,

    @Lob
    var plcySprtCn: String? = null,

    var sprvsnInstCdNm: String? = null,
    var operInstCdNm: String? = null,
    var aplyPrdSeCd: String? = null,
    var bizPrdBgngYmd: String? = null,
    var bizPrdEndYmd: String? = null,

    @Lob
    var plcyAplyMthdCn: String? = null,

    @Lob
    var aplyUrlAddr: String? = null,

    @Lob
    var sbmsnDcmntCn: String? = null,

    var sprtTrgtMinAge: String? = null,
    var sprtTrgtMaxAge: String? = null,
    var sprtTrgtAgeLmtYn: String? = null,
    var mrgSttsCd: String? = null,
    var earnCndSeCd: String? = null,
    var earnMinAmt: String? = null,
    var earnMaxAmt: String? = null,

    @Lob
    var zipCd: String? = null,

    var jobCd: String? = null,
    var schoolCd: String? = null,
    var aplyYmd: String? = null,
    var sBizCd: String? = null,

    @Lob
    @Column(columnDefinition = "TEXT")
    var rawJson: String? = null
) {
    companion object {
        @JvmStatic
        fun builder(): PolicyBuilder = PolicyBuilder()

        @JvmStatic
        fun from(item: PolicyItem, rawJson: String?): Policy {
            return Policy(
                plcyNo = item.plcyNo,
                plcyNm = item.plcyNm,
                plcyKywdNm = item.plcyKywdNm,
                plcyExplnCn = item.plcyExplnCn,
                plcySprtCn = item.plcySprtCn,
                sprvsnInstCdNm = item.sprvsnInstCdNm,
                operInstCdNm = item.operInstCdNm,
                aplyPrdSeCd = item.aplyPrdSeCd,
                bizPrdBgngYmd = item.bizPrdBgngYmd,
                bizPrdEndYmd = item.bizPrdEndYmd,
                plcyAplyMthdCn = item.plcyAplyMthdCn,
                aplyUrlAddr = item.aplyUrlAddr,
                sbmsnDcmntCn = item.sbmsnDcmntCn,
                sprtTrgtMinAge = item.sprtTrgtMinAge,
                sprtTrgtMaxAge = item.sprtTrgtMaxAge,
                sprtTrgtAgeLmtYn = item.sprtTrgtAgeLmtYn,
                mrgSttsCd = item.mrgSttsCd,
                earnCndSeCd = item.earnCndSeCd,
                earnMinAmt = item.earnMinAmt,
                earnMaxAmt = item.earnMaxAmt,
                zipCd = item.zipCd,
                jobCd = item.jobCd,
                schoolCd = item.schoolCd,
                aplyYmd = item.aplyYmd,
                sBizCd = item.sbizCd,
                rawJson = rawJson
            )
        }
    }
}

class PolicyBuilder {
    var plcyNo: String? = null
    var plcyNm: String? = null
    var plcyKywdNm: String? = null
    var plcyExplnCn: String? = null
    var plcySprtCn: String? = null
    var sprvsnInstCdNm: String? = null
    var operInstCdNm: String? = null
    var aplyPrdSeCd: String? = null
    var bizPrdBgngYmd: String? = null
    var bizPrdEndYmd: String? = null
    var plcyAplyMthdCn: String? = null
    var aplyUrlAddr: String? = null
    var sbmsnDcmntCn: String? = null
    var sprtTrgtMinAge: String? = null
    var sprtTrgtMaxAge: String? = null
    var sprtTrgtAgeLmtYn: String? = null
    var mrgSttsCd: String? = null
    var earnCndSeCd: String? = null
    var earnMinAmt: String? = null
    var earnMaxAmt: String? = null
    var zipCd: String? = null
    var jobCd: String? = null
    var schoolCd: String? = null
    var aplyYmd: String? = null
    var sBizCd: String? = null
    var rawJson: String? = null

    fun plcyNo(v: String?) = apply { plcyNo = v }
    fun plcyNm(v: String?) = apply { plcyNm = v }
    fun plcyKywdNm(v: String?) = apply { plcyKywdNm = v }
    fun plcyExplnCn(v: String?) = apply { plcyExplnCn = v }
    fun plcySprtCn(v: String?) = apply { plcySprtCn = v }
    fun sprvsnInstCdNm(v: String?) = apply { sprvsnInstCdNm = v }
    fun operInstCdNm(v: String?) = apply { operInstCdNm = v }
    fun aplyPrdSeCd(v: String?) = apply { aplyPrdSeCd = v }
    fun bizPrdBgngYmd(v: String?) = apply { bizPrdBgngYmd = v }
    fun bizPrdEndYmd(v: String?) = apply { bizPrdEndYmd = v }
    fun plcyAplyMthdCn(v: String?) = apply { plcyAplyMthdCn = v }
    fun aplyUrlAddr(v: String?) = apply { aplyUrlAddr = v }
    fun sbmsnDcmntCn(v: String?) = apply { sbmsnDcmntCn = v }
    fun sprtTrgtMinAge(v: String?) = apply { sprtTrgtMinAge = v }
    fun sprtTrgtMaxAge(v: String?) = apply { sprtTrgtMaxAge = v }
    fun sprtTrgtAgeLmtYn(v: String?) = apply { sprtTrgtAgeLmtYn = v }
    fun mrgSttsCd(v: String?) = apply { mrgSttsCd = v }
    fun earnCndSeCd(v: String?) = apply { earnCndSeCd = v }
    fun earnMinAmt(v: String?) = apply { earnMinAmt = v }
    fun earnMaxAmt(v: String?) = apply { earnMaxAmt = v }
    fun zipCd(v: String?) = apply { zipCd = v }
    fun jobCd(v: String?) = apply { jobCd = v }
    fun schoolCd(v: String?) = apply { schoolCd = v }
    fun aplyYmd(v: String?) = apply { aplyYmd = v }
    fun sBizCd(v: String?) = apply { sBizCd = v }
    fun rawJson(v: String?) = apply { rawJson = v }

    fun build(): Policy = Policy(
        plcyNo = plcyNo,
        plcyNm = plcyNm,
        plcyKywdNm = plcyKywdNm,
        plcyExplnCn = plcyExplnCn,
        plcySprtCn = plcySprtCn,
        sprvsnInstCdNm = sprvsnInstCdNm,
        operInstCdNm = operInstCdNm,
        aplyPrdSeCd = aplyPrdSeCd,
        bizPrdBgngYmd = bizPrdBgngYmd,
        bizPrdEndYmd = bizPrdEndYmd,
        plcyAplyMthdCn = plcyAplyMthdCn,
        aplyUrlAddr = aplyUrlAddr,
        sbmsnDcmntCn = sbmsnDcmntCn,
        sprtTrgtMinAge = sprtTrgtMinAge,
        sprtTrgtMaxAge = sprtTrgtMaxAge,
        sprtTrgtAgeLmtYn = sprtTrgtAgeLmtYn,
        mrgSttsCd = mrgSttsCd,
        earnCndSeCd = earnCndSeCd,
        earnMinAmt = earnMinAmt,
        earnMaxAmt = earnMaxAmt,
        zipCd = zipCd,
        jobCd = jobCd,
        schoolCd = schoolCd,
        aplyYmd = aplyYmd,
        sBizCd = sBizCd,
        rawJson = rawJson
    )
}

File: frontend/src/types/policy.ts
// Policy 관련 타입 정의
// 백엔드 API 계약 기반
// GET /api/v1/welfare/policy/search

// ===== 검색 요청 파라미터 =====
// PolicyElasticSearchRequestDto 기반
export interface PolicySearchRequest {
  zipCd?: string | null;
  schoolCode?: string | null;
  jobCode?: string | null;
  keyword?: string | null;
  age?: number | null;
  earn?: number | null;
  regionCode?: string | null;
  marriageStatus?: string | null;
  keywords?: string[] | null;
  from: number;
  size: number;
}

// ===== 검색 드롭다운용 코드 맵 =====
// 백엔드 CodeEnum 기반 (code → 한글 라벨)

// EducationRequirement (학력요건)
export const SchoolCodeLabel: Record<string, string> = {
  "049001": "고졸 미만",
  "049002": "고교 재학",
  "049003": "고졸 예정",
  "049004": "고교 졸업",
  "049005": "대학 재학",
  "049006": "대졸 예정",
  "049007": "대학 졸업",
  "049008": "석·박사",
  "049009": "기타",
  "049010": "제한없음",
};

// JobRequirement (취업요건)
export const JobCodeLabel: Record<string, string> = {
  "013001": "재직자",
  "013002": "자영업자",
  "013003": "미취업자",
  "013004": "프리랜서",
  "013005": "일용근로자",
  "013006": "(예비)창업자",
  "013007": "단기근로자",
  "013008": "영농종사자",
  "013009": "기타",
  "013010": "제한없음",
};

// MarriageRequirement (결혼상태)
export const MarriageStatusCodeLabel: Record<string, string> = {
  "055001": "기혼",
  "055002": "미혼",
  "055003": "제한없음",
};

// ===== 검색 응답 =====
// PolicyDocument 기반
export interface PolicyDocument {
  policyId: number | null;
  plcyNo: string | null;
  plcyNm: string | null;

  // 나이
  minAge: number | null;
  maxAge: number | null;
  ageLimited: boolean | null;

  // 소득
  earnCondition: string | null;
  earnMin: number | null;
  earnMax: number | null;

  // 대상 조건
  regionCode: string | null;
  jobCode: string | null;
  schoolCode: string | null;
  marriageStatus: string | null;

  // 태그 / 분류
  keywords: string[] | null;
  specialBizCode: string | null;

  // 검색용 텍스트
  description: string | null;
}

// ===== 북마크 응답 =====
// GET /api/v1/member/bookmark/welfare-bookmarks
// BookmarkPolicyResponseDto 기반
export interface BookmarkPolicyResponse {
  code: number;
  message: string;
  policies: Policy[] | null;
}

// ===== 북마크 토글 응답 =====
// POST /api/v1/member/bookmark/welfare-bookmarks/{policyId}
// BookmarkUpdateResponseDto 기반
export interface BookmarkUpdateResponse {
  code: number;
  message: string;
}

// ===== 신청 관련 =====
// GET /api/v1/member/policy-aply/welfare-applications
// Application 엔티티 그대로 직렬화 (Policy, Member 포함)
export interface ApplicationItem {
  id: number;
  policy: Policy;
  applicant: ApplicationMember;
  createdAt: string; // ISO 8601
  modifiedAt: string; // ISO 8601
}

// Application 내 Member (엔티티 전체 직렬화)
export interface ApplicationMember {
  id: number;
  name: string;
  email: string;
}

// POST /api/v1/member/policy-aply/welfare-application/{policyId}
// AddApplicationResponseDto 기반
export interface AddApplicationResponse {
  status: number;
  message: string;
}

// PUT /api/v1/member/policy-aply/welfare-application/{id}
// DeleteApplicationResponseDto 기반
export interface DeleteApplicationResponse {
  code: number;
  message: string;
}

// Policy 엔티티 기반
export interface Policy {
  id: number | null;
  plcyNo: string | null;
  plcyNm: string | null;
  plcyKywdNm: string | null;
  plcyExplnCn: string | null;
  plcySprtCn: string | null;
  sprvsnInstCdNm: string | null;
  operInstCdNm: string | null;
  aplyPrdSeCd: string | null;
  bizPrdBgngYmd: string | null;
  bizPrdEndYmd: string | null;
  plcyAplyMthdCn: string | null;
  aplyUrlAddr: string | null;
  sbmsnDcmntCn: string | null;
  sprtTrgtMinAge: string | null;
  sprtTrgtMaxAge: string | null;
  sprtTrgtAgeLmtYn: string | null;
  mrgSttsCd: string | null;
  earnCndSeCd: string | null;
  earnMinAmt: string | null;
  earnMaxAmt: string | null;
  zipCd: string | null;
  jobCd: string | null;
  schoolCd: string | null;
  aplyYmd: string | null;
  sbizCd: string | null;
  rawJson: string | null;
}
